db2-contacts.js

```diff
@@ -0,0 +1,195 @@
+  // to help the poor bugger
+  let batchKeys = {} // key to index
+
+  function writeData(cb) {
+    level.batch(batch, { valueEncoding: 'json' }, cb)
+    batch = []
+    batchKeys = {}
+  }
+
+  function handleData(record, processed) {
+    if (record.offset < offset.value) return batch.length
+    const recBuffer = record.value
+    if (!recBuffer) return batch.length // deleted
+
+    let p = 0 // note you pass in p!
+    p = bipf.seekKey(recBuffer, p, bValue)
+    if (!~p) return batch.length
+
+    const pAuthor = bipf.seekKey(recBuffer, p, bAuthor)
+    const author = bipf.decode(recBuffer, pAuthor)
+
+    const pContent = bipf.seekKey(recBuffer, p, bContent)
+    if (!~pContent) return batch.length
+
+    const pType = bipf.seekKey(recBuffer, pContent, bType)
+    if (!~pType) return batch.length
+
+    if (bipf.compareString(recBuffer, pType, bContact) === 0) {
+      const content = bipf.decode(recBuffer, pContent)
+      const to = content.contact
+
+      if (isFeed(author) && isFeed(to)) {
+        const value = content.blocking || content.flagged ? -1 :
+              content.following === true ? 1
+              : -2
+
+        let updateFeeds = false
+
+        let fromIndex = feeds.indexOf(author)
+        if (fromIndex === -1) {
+          feeds.push(author)
+          fromIndex = feeds.length -1
+          updateFeeds = true
+        }
+
+        let toIndex = feeds.indexOf(to)
+        if (toIndex === -1) {
+          feeds.push(to)
+          toIndex = feeds.length -1
+          updateFeeds = true
+        }
+
+        let fromValues = feedValues[fromIndex] || {}
+        fromValues[toIndex] = value
+        feedValues[fromIndex] = fromValues
+
+        const batchValue = {
+          type: 'put',
+          key: fromIndex,
+          value: fromValues
+        }
+
+        let existingKeyIndex = batchKeys[fromIndex]
+        if (existingKeyIndex) {
+          batch[existingKeyIndex] = batchValue
+        }
+        else {
+          batch.push(batchValue)
+          batchKeys[fromIndex] = batch.length - 1
+        }
+
+        if (updateFeeds) {
+          const feedsValue = {
+            type: 'put',
+            key: 'feeds',
+            value: feeds
+          }
+
+          let existingFeedsIndex = batchKeys['feeds']
+          if (existingFeedsIndex) {
+            batch[existingFeedsIndex] = feedsValue
+          } else {
+            batch.push(feedsValue)
+            batchKeys['feeds'] = batch.length - 1
+          }
+        }
+
+        layer(author, to, value)
+      }
+    }
+
+    return batch.length
+  }
+
+  layer({})
+
+  function beforeIndexUpdate(cb) {
+    get((err, g) => {
+      layer(g)
+      cb()
+    })
+  }
+
+  function get(cb) {
+    pull(
+      pl.read(level, {
+        valueEncoding: jsonCodec,
+        keys: true
+      }),
+      pull.collect((err, data) => {
+        if (err) return cb(err)
+
+        for (let i = 0; i < data.length; ++i) {
```

Comment:
Reading this backwards might be a speed up, because level scans the rows in lexicographic order of the key, and we have 99% of the keys are numbers, so I think `'feed'` will be the very last one. Not sure if it helps, but maybe worth putting a console.log inside the for-loop and then see whether doing it forwards or backwards is the best.

Reply:
I just tried. Didn't really change anything. The part from collect to cb is 14ms on my machine.